src/cdtools/models/base.py

```
@@ -521,177 +374,202 @@ def Adam_optimize(
+        regularization_factor : float or list(float).
+            Optional, if the model has a regularizer defined, the set of 
+            parameters to pass the regularizer method.
         thread : bool
-            Default True, whether to run the computation in a separate thread to allow interaction with plots during computation
+            Default True, whether to run the computation in a separate thread 
+            to allow interaction with plots during computation.
         calculation_width : int
-            Default 10, how many translations to pass through at once for each round of gradient accumulation. Does not affect the result, only the calculation speed
-
+            Default 10, how many translations to pass through at once for 
+            each round of gradient accumulation. Does not affect the result, 
+            only the calculation speed. 
+        
         """
-
-        self.training_history += (
-            f'Planning {iterations} epochs of Adam, with a learning rate = '
-            f'{lr}, batch size = {batch_size}, regularization_factor = '
-            f'{regularization_factor}, and schedule = {schedule}.\n'
-        )
+        # We want to have model.Adam_optimize call AND store cdtools.reconstructors.Adam
+        # to perform reconstructions without creating a new reconstructor each time we 
+        # update the hyperparameters.
+        # 
+        # The only way to do this is to make the Adam reconstructor an attribute
+        # of the model. But since the Adam reconstructor also depends on CDIModel,
+        # this seems to give rise to a circular import error unless
+        # we import cdtools.reconstructors within this method:
+        if not hasattr(self, 'reconstructor'):
+            from cdtools.reconstructors import Adam
+            self.reconstructor = Adam(model=self, 
+                                  dataset=dataset, 
+                                  subset=subset)
```

This needs to be updated so that if the user changes which optimizer they're using, it doesn't pass a silent error. I would lean towards just creating a new optimizer for each new function (less good behavior, but preserving old behavior for old scripts), but I could be swayed for sure. What do other folks think?